File: src/main/scala/comptime/hir.scala
// package comptime

// import vfs.VfsNode
// import parse.AstNode

// trait Attribute {
//   var attributes: Map[String, Value] = Map()
//   def getAttribute(name: String): Option[Value] = attributes.get(name)

//   def withAttribute(name: String, value: Value): this.type = {
//     attributes += (name -> value)
//     this
//   }

//   def withChangeSet(new_attributes: Map[String, Value]): this.type = {
//     attributes = attributes ++ new_attributes
//     this
//   }
// }

// trait Symbol {
//   var name: Option[String] = None
//   def getName: Option[String] = name
//   def withName(newName: String): this.type = {
//     name = Some(newName)
//     this
//   }
// }

// trait ASTTrackable {
//   var ast: Option[AstLocation] = None
//   def getAstNode: Option[AstLocation] = ast
//   def withAstNode(newAst: AstLocation): this.type = {
//     ast = Some(newAst)
//     this
//   }
// }

// trait Namespace {
//   var children: Map[String, Value] = Map()
//   var uses: List[UseImport] = List()
//   // 检查命名空间中的符号
//   def getChildren: Map[String, Value] = children
//   def withChildren(newChildren: Map[String, Value]): this.type = {
//     children = newChildren
//     this
//   }
//   def withChild(name: String, value: Value): this.type = {
//     children += (name -> value)
//     this
//   }
//   def withMoreChildren(newChildren: Map[String, Value]): this.type = {
//     children = children ++ newChildren
//     this
//   }
//   def withUseImport(use: UseImport): this.type = {
//     uses = use :: uses
//     this
//   }
//   def withUseImports(newUses: List[UseImport]): this.type = {
//     uses = newUses ++ uses
//     this
//   }
//   def removeChild(name: String): Unit = children -= name
//   def removeChildren(names: List[String]): Unit = names.foreach(removeChild)
//   def removeUseImport(use: UseImport): Unit = uses = uses.filterNot(_ == use)
//   def removeUseImports(newUses: List[UseImport]): Unit = uses = uses.filterNot(newUses.contains)
//   def removeAllUseImports(): Unit = uses = List()

//   def lookup(name: String): Option[Value] = children.get(name) match
//     case Some(value) => Some(value)
//     case None => uses.lookup(name)

//   def lookup(path: List[String]): Option[Value] =
//     var result: Option[Value] = None
//     var index = 0
//     while (index < path.length && result.isEmpty) {
//       val name = path(index)
//       result = lookup(name)
//       index += 1
//     }
//     result
// }

// enum UseImport {
//   case UseAll(namespace: Definition)
//   case UseAllExcept(namespace: Definition, except: List[String])
//   case UseOnly(namespace: Definition, only: List[String])
//   case UseThis(namespace: Definition, thisName: String, asName: String)
// }

// extension (uses: List[UseImport]) {
//   def lookup(name: String): Option[Value] = uses.flatMap {
//     case UseImport.UseAll(namespace) => namespace.lookup(name)
//     case UseImport.UseAllExcept(namespace, except) if !except.contains(name) => namespace.lookup(name)
//     case UseImport.UseOnly(namespace, only) if only.contains(name) => namespace.lookup(name)
//     case UseImport.UseThis(namespace, thisName, asName) if asName == name => namespace.lookup(thisName)
//     case _ => None
//   }.headOption
// }

// trait Parent {
//   var parent: Option[Definition] = None;

//   def getParent: Option[Definition] = parent
//   def withParent(newParent: Definition): this.type = {
//     parent = Some(newParent)
//     this
//   }
// }

// trait Definition extends Namespace, Parent, Symbol, Value {
//   def dumpNamespace(): String =
//     val cs = getChildren.map { case (k, v) => s"${v.asInstanceOf[Definition].dumpNamespace()}" }.mkString(" ")
//     val childrenStr = s"(children $cs)"
//     s"(${getName.getOrElse("unknown")} ${this.getClass().getName()} $childrenStr)"

//   def dumpNamespace(writer: java.io.PrintWriter): Unit = writer.write(dumpNamespace())

//   // 命名空间中解析符号
//   def resolve(name: String): Option[Value] = lookup(name) match
//     case Some(value) => Some(value)
//     case None => getParent.flatMap(_.resolve(name))

//   def absolutePath: List[String] = Nil
// }

// enum Param {
//   case Simple(id: String, ty: Type, attributes: Map[String, Value] = Map())
//   case Named(id: String, ty: Type, default: Value, attributes: Map[String, Value] = Map())
//   case Vararg(id: String, ty: Type, attributes: Map[String, Value] = Map())
//   case Comptime(id: String, ty: Type, attributes: Map[String, Value] = Map())
//   case Implicit(id: String, ty: Type, attributes: Map[String, Value] = Map())
//   case Self(id: String, attributes: Map[String, Value] = Map())
//   case RefSelf(id: String, attributes: Map[String, Value] = Map())
//   case Itself(id: String, attributes: Map[String, Value] = Map())
//   case RefItself(id: String, attributes: Map[String, Value] = Map())
//   case Id(id: String, attributes: Map[String, Value] = Map())
// }

// enum Clause {
//   case TypeDeclaration(id: String)
//   case TypedDeclaration(id: String, ty: Type, attributes: Map[String, Value] = Map())
//   case TraitBound(id: String, trait_bound: Value, attributes: Map[String, Value] = Map())
//   // TODO
//   case Requires()
//   case Ensures()
//   case Outcomes()
// }

// trait Expr(ast: Option[AstLocation] = None) extends Value {
//   def eval(): ExprEvalResult = throw new NotImplementedError("eval() not implemented")
// }

// enum ExprEvalResult {
//   case Comptime(value: Value)
//   case Runtime(expr: Expr)
// }

// trait Pattern(yields: List[Declaration] = List(), ast: Option[AstLocation] = None) extends Value

// trait Statement(attributes: Map[String, Value] = Map(), ast: Option[AstLocation] = None) extends Expr {
//   override def getType: Type = VoidType()
// }

// case class AstLocation(file: VfsNode, node: AstNode)

File: src/main/kotlin/comptime/hir.kt
package comptime

import vfs.VfsNode
import parse.AstNode

interface Attribute {
    var attributes: MutableMap<String, Value>

    fun getAttribute(name: String): Value? = attributes[name]

    fun setAttribute(name: String, value: Value) {
        attributes[name] = value
    }

    fun updateChangeSet(newAttributes: Map<String, Value>) {
        attributes.putAll(newAttributes)
    }
}

interface Symbol {
    var name: String?

    fun getName(): String? = name
    fun setName(newName: String) {
        name = newName
    }
}

interface ASTTrackable {
    var ast: AstLocation?

    fun getAst(): AstLocation? = ast
    fun setAst(newAst: AstLocation) {
        ast = newAst
    }
}

interface Namespace {
    var children: MutableMap<String, Value>
    var uses: MutableList<UseImport>

    // 检查命名空间中的符号
    fun getChildren(): Map<String, Value> = children
    fun setChildren(newChildren: Map<String, Value>) {
        children = newChildren.toMutableMap()
    }
    fun addChild(name: String, value: Value) {
        children[name] = value
    }
    fun addChildren(newChildren: Map<String, Value>) {
        children.putAll(newChildren)
    }
    fun addUseImport(use: UseImport) {
        uses.add(0, use)
    }
    fun addUseImports(newUses: List<UseImport>) {
        uses.addAll(0, newUses)
    }
    fun removeChild(name: String) {
        children.remove(name)
    }
    fun removeChildren(names: List<String>) {
        names.forEach { removeChild(it) }
    }
    fun removeUseImport(use: UseImport) {
        uses.removeAll { it == use }
    }
    fun removeUseImports(newUses: List<UseImport>) {
        uses.removeAll { newUses.contains(it) }
    }
    fun removeAllUseImports() {
        uses.clear()
    }

    fun lookup(name: String): Value? = children[name] ?: uses.lookup(name)

    fun lookup(path: List<String>): Value? {
        var result: Value? = null
        var index = 0
        while (index < path.size && result == null) {
            val name = path[index]
            result = lookup(name)
            index++
        }
        return result
    }
}

sealed class UseImport {
    class UseAll(val namespace: Definition) : UseImport()
    class UseAllExcept(val namespace: Definition, val except: List<String>) : UseImport()
    class UseOnly(val namespace: Definition, val only: List<String>) : UseImport()
    class UseThis(val namespace: Definition, val thisName: String, val asName: String) : UseImport()
}

// Kotlin扩展函数替代Scala的extension
fun List<UseImport>.lookup(name: String): Value? {
    return this.firstNotNullOfOrNull { use ->
        when (use) {
            is UseImport.UseAll -> use.namespace.lookup(name)
            is UseImport.UseAllExcept -> if (!use.except.contains(name)) use.namespace.lookup(name) else null
            is UseImport.UseOnly -> if (use.only.contains(name)) use.namespace.lookup(name) else null
            is UseImport.UseThis -> if (use.asName == name) use.namespace.lookup(use.thisName) else null
        }
    }
}

interface Parent {
    var parent: Definition?

    fun getParent(): Definition? = parent
    fun setParent(newParent: Definition) {
        parent = newParent
    }
}

interface Definition : Namespace, Parent, Symbol, Value {
    fun dumpNamespace(): String {
        println("Dumping namespace for ${this.javaClass.name}")
        val cs = getChildren().entries.joinToString(" ") { (k, v) ->
            "($k ${v.javaClass.name} ${(v as Definition).dumpNamespace()})"
        }
        val childrenStr = "(children $cs)"
        return "(${getName() ?: "unknown"} $childrenStr)"
    }

    fun dumpNamespace(writer: java.io.PrintWriter) {
        writer.write(dumpNamespace())
    }

    // 命名空间中解析符号
    fun resolve(name: String): Value? = lookup(name) ?: getParent()?.resolve(name)

    fun absolutePath(): List<String> = emptyList()
}

sealed class Param {
    class Simple(val id: String, val ty: Type, val attributes: Map<String, Value> = emptyMap()) : Param()
    class Named(val id: String, val ty: Type, val default: Value, val attributes: Map<String, Value> = emptyMap()) : Param()
    class Vararg(val id: String, val ty: Type, val attributes: Map<String, Value> = emptyMap()) : Param()
    class Comptime(val id: String, val ty: Type, val attributes: Map<String, Value> = emptyMap()) : Param()
    class Implicit(val id: String, val ty: Type, val attributes: Map<String, Value> = emptyMap()) : Param()
    class Self(val id: String, val attributes: Map<String, Value> = emptyMap()) : Param()
    class RefSelf(val id: String, val attributes: Map<String, Value> = emptyMap()) : Param()
    class Itself(val id: String, val attributes: Map<String, Value> = emptyMap()) : Param()
    class RefItself(val id: String, val attributes: Map<String, Value> = emptyMap()) : Param()
    class Id(val id: String, val attributes: Map<String, Value> = emptyMap()) : Param()
}

sealed class Clause {
    class TypeDeclaration(val id: String) : Clause()
    class TypedDeclaration(val id: String, val ty: Type, val attributes: Map<String, Value> = emptyMap()) : Clause()
    class TraitBound(val id: String, val traitBound: Value, val attributes: Map<String, Value> = emptyMap()) : Clause()
    // TODO
    class Requires : Clause()
    class Ensures : Clause()
    class Outcomes : Clause()
}

// 在Kotlin中，抽象类更接近Scala的traits
abstract class Expr(open val ast: AstLocation? = null) : Value

abstract class Pattern(val yields: List<Declaration> = emptyList(), override val ast: AstLocation? = null) : Value

abstract class Statement(val attributes: Map<String, Value> = emptyMap(), override val ast: AstLocation? = null) : Expr(ast) {
    override fun getType(): Type = VoidType()
}

data class AstLocation(val file: VfsNode, val node: AstNode)